### PR TITLE
Implement Array#first(n : Int) and Array#last(n : Int)

### PR DESCRIPTION
**Pull request by greyblake**
Implement `Array#first(n : Int)` and `Array#last(n : Int)` methods.
E.g.
```crystal
[1, 2, 3].first(2) # => [1, 2]
[1, 2, 3].first(4) # => [1, 2, 3]

[1, 2, 3].last(2) # => [2, 3]
[1, 2, 3].last(4) # => [1, 2, 3]
```

### CODE REVIEW

**Comment by jhass** [comment]
I think we had `#first(n)` and dropped it in favor of `#take(n)` to remove redundancy.

**Comment by greyblake** [comment]
Oh.. It's not so obvious. `#take` comes from Enumerable. Maybe, there should be an alias for `first`?

Still, is there something similar to `#last?`

File: src/array.cr [hunk 1]
```diff
@@ -997,6 +997,16 @@ class Array(T)
     @size == 0 ? yield : @buffer[0]
   end
 
+  # Returns first `n` elements of the array.
```
Comment (kirbyfan64) [inline src/array.cr]: Returns **the** first

File: src/array.cr [hunk 2]
```diff
@@ -1069,6 +1079,20 @@ class Array(T)
     @size == 0 ? yield : @buffer[@size - 1]
   end
 
+  # Returns last `n` elements of the array.
```
Comment (kirbyfan64) [inline src/array.cr]: Same here.

File: src/array.cr (hunk 1 above)
Comment (greyblake) [inline src/array.cr]: addressed in bb65c05

File: src/array.cr (hunk 2 above)
Comment (greyblake) [inline src/array.cr]: addressed in bb65c05

**Comment by asterite** [comment]
I didn't know there was `Array#last(n)` in Ruby, it looks nice and useful. But, if we add it, then it would also make sense to have `Array#first(n)`. We also have `Enumerable#take(n)`. I chose `take` instead of `first` because it's also defined in `Iterator`, and there I think it maybe makes more sense to say "take" instead of "first", but now I'm not sure. The docs say "Returns an iterator that only returns the **first n** elements of the initial iterator".

Would you (community) get very mad if we use `first` and remove `take`? I know in a [previous commit](https://github.com/crystal-lang/crystal/commit/0ac2c2abe55e7ad3d4be8060007f48c9bfaeab9f) we removed `first` in favor of `take`, but I know think that was a mistake and we should have removed `take` and kept `first`.  Or we could have both, but we prefer to avoid having aliases. Also with `take` it isn't very clear that it returns the *first* elements. Alternatively, `take(n)` in an iterator would return an iterator, but `first(n)` would return an Array, but maybe that's confusing.

So my vote is to rename `take` to `first` and also add `last(n)`. There's also #1896 that suggests adding `Enumerable#last`, we could look at that (I remember @waj was in favor of having such method).

**Comment by jhass** [comment]
+1 I prefer `first` over `take` in Ruby too.

**Comment by greyblake** [comment]
+1 for `take(n) -> first(n)` + `last(n)`

P.S. I love this way of solving the problems! :)

**Comment by ozra** [comment]
First of all, I'm all for having both, but having to chose, I'd take `first` (aurrggh, so not funny, I must get some sleep.). Well +1 `first` (and `last`).

**Comment by bcardiff** [comment]
[Haskell fan here!]
I prefer `take` because it's dual with `drop`
[/Haskell fan here!]

Yet, I think I will lost this one :-).

**Comment by asterite** [comment]
I'll change `take` to `first` then :-)

@greyblake Thanks!

**Comment by asterite** [comment]
@bcardiff Sorry, but if we keep `take(n)` then we have to remove `first(n)` if we don't want to have aliases. But we already have `first()` and `last()` (for Array), so it makes more sense to keep `first` + `last` , fits better into the overall set of methods.

**Comment by greyblake** [comment]
Yahoo)) Thanks!:)

**Pull request merged**